artichoke-backend/src/extn/core/time/trampoline.rs

```diff
@@ -1,35 +1,110 @@
-        0_i64
-    };
+pub fn at(
+    interp: &mut Artichoke,
+    seconds: Value,
+    first: Option<Value>,
+    second: Option<Value>,
+    third: Option<Value>,
+) -> Result<Value, Error> {
+    // Coerce the params to the correct place. Specifically:
+    // - the options hash might not always be provided as the last argument.
+    // - subseconds can be provided with an optional symbol for the type of subsec.
+    //
+    // ```console
+    // [3.1.2] > Time.at(1)
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1)
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, in: "A")
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1, in: "A")
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec, in: "A")
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // ```
 
-    if let Some(time) = Time::at(seconds, sub_second_nanos) {
-        let result = Time::alloc_value(time, interp)?;
-        Ok(result)
+    let mut subsec = first;
+    let mut subsec_unit = second;
+    let mut options = third;
+
+    // Re-position the options hash under the `options` if it exists. Calling
+    // `Time.at` without the optional parameters will end up placing the
+    // options hash in the incorrect parameter position.
+    //
+    // ```console
+    // Time.at(0, in: "A")
+    // #          ^--first
+    // Time.at(0, 1, in: "A")
+    // #             ^-- second
+    // Time.at(0, 1, :nsec, in: "A")
+    // #                    ^-- third
+    // ```
+    //
+    // The below logic:
+    // - ensures the third parameter is a Ruby::Hash if provided.
+    // - if third param is not options, check the second paramter, if it is a
+    //   Ruby::Hash then assume this is the options hash, and clear out the
+    //   second parameter.
+    // - if second param is not options, check the first param, if it is a
+    //   Ruby::Hash then assume this is the options hash, and clear out the
+    //   first parameter.
+    if let Some(third_param) = third {
+        if third_param.ruby_type() != Ruby::Hash {
+            return Err(ArgumentError::with_message("invalid offset options").into());
+        }
     } else {
-        Err(ArgumentError::with_message("Time too large").into())
+        options = if let Some(second_param) = second {
+            if second_param.ruby_type() == Ruby::Hash {
+                subsec_unit = None;
+                Some(second_param)
+            } else if let Some(first_param) = first {
+                if first_param.ruby_type() == Ruby::Hash {
+                    subsec = None;
+                    Some(first_param)
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        }
```

Comment:
wow this is a gnarly expression! I don't see a better way to do this though.

Reply:
Oh good! I was like "really, is this it? There must be a better way". I'm sure there some array plus iterator magic some code golfer could come up with hah.

Reply:
hopefully once `if let ... else` lands this can get cleaned up a bunch.